File: src/code/python/data_structures/binary_search_tree.py
from typing import Any


class TreeNode:
    def __init__(self, data: Any) -> None:
        self.data = data
        self.left = None
        self.right = None


class BinarySearchTree:
    def __init__(self) -> None:
        self.root = None

    def insert(self, data: Any) -> None:
        if not self.root:
            self.root = TreeNode(data)
        else:
            self.insert_node(self.root, data)

    def insert_node(self, node: TreeNode | None, data: Any) -> None:
        if data < node.data:
            if not node.left:
                node.left = TreeNode(data)
            else:
                self.insert_node(node.left, data)
        else:
            if not node.right:
                node.right = TreeNode(data)
            else:
                self.insert_node(node.right, data)

    def __repr__(self) -> str:
        return 'Empty tree' if not self.root else self._print_tree(self.root, '', True)

    def _print_tree(self, node: TreeNode | None, prefix: str, is_left: bool) -> str:
        if node is None:
            return ''

        result = ''
        result += self._print_tree(node.right, prefix + ('│   ' if is_left else '    '), False)
        result += prefix + ('└── ' if is_left else '┌── ') + str(node.data) + '\n'
        result += self._print_tree(node.left, prefix + ('    ' if is_left else '│   '), True)

        return result

File: src/code/ruby/data_structures/binary_search_tree.rb
class TreeNode
  attr_accessor :data, :left, :right

  def initialize(data)
    @data   = data
    @left   = nil
    @right  = nil
  end
end

class BinarySearchTree
  attr_accessor :root

  def initialize
    @root = nil
  end

  def insert(data)
    if @root.nil?
      @root = TreeNode.new(data)
    else
      insert_node(@root, data)
    end
  end

  def insert_node(node, data)
    if data < node.data
      if node.left.nil?
        node.left = TreeNode.new(data)
      else
        insert_node(node.left, data)
      end
    else
      if node.right.nil?
        node.right = TreeNode.new(data)
      else
        insert_node(node.right, data)
      end
    end
  end

  def to_s
    @root.nil? ? 'Empty tree' : print_tree(@root, '', true)
  end

  private

  def print_tree(node, prefix, is_left)
    return '' if node.nil?

    result = ''
    result += print_tree(node.right, prefix + (is_left ? '│   ' : '    '), false)
    result += prefix + (is_left ? '└── ' : '┌── ') + node.data.to_s + "\n"
    result += print_tree(node.left, prefix + (is_left ? '    ' : '│   '), true)

    result
  end
end
